airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/AbstractDbSource.java

```diff
@@ -135,36 +135,49 @@ public AirbyteCatalog discover(final JsonNode config) throws Exception {
-            .collect(Collectors.toMap(t -> String.format("%s.%s", t.getNameSpace(), t.getName()), Function
-                .identity()));
-
-    validateCursorFieldForIncrementalTables(fullyQualifiedTableNameToInfo, catalog);
-
-    final List<AutoCloseableIterator<AirbyteMessage>> incrementalIterators =
-        getIncrementalIterators(database, catalog, fullyQualifiedTableNameToInfo, stateManager, emittedAt);
-    final List<AutoCloseableIterator<AirbyteMessage>> fullRefreshIterators =
-        getFullRefreshIterators(database, catalog, fullyQualifiedTableNameToInfo, stateManager, emittedAt);
-    final List<AutoCloseableIterator<AirbyteMessage>> iteratorList = Stream
-        .of(incrementalIterators, fullRefreshIterators)
-        .flatMap(Collection::stream)
-        .collect(Collectors.toList());
+    throws Exception {
+    try {
+      final StateManager stateManager =
+          StateManagerFactory.createStateManager(getSupportedStateType(config), deserializeInitialState(state, config), catalog);
+      final Instant emittedAt = Instant.now();
 
-    return AutoCloseableIterators
-        .appendOnClose(AutoCloseableIterators.concatWithEagerClose(iteratorList), () -> {
-          LOGGER.info("Closing database connection pool.");
-          Exceptions.toRuntime(this::close);
-          LOGGER.info("Closed database connection pool.");
-        });
+      final Database database = createDatabaseInternal(config);
+
+      final Map<String, TableInfo<CommonField<DataType>>> fullyQualifiedTableNameToInfo =
+          discoverWithoutSystemTables(database)
+              .stream()
+              .collect(Collectors.toMap(t -> String.format("%s.%s", t.getNameSpace(), t.getName()), Function
+                  .identity()));
+
+      validateCursorFieldForIncrementalTables(fullyQualifiedTableNameToInfo, catalog);
+
+      final List<AutoCloseableIterator<AirbyteMessage>> incrementalIterators =
+          getIncrementalIterators(database, catalog, fullyQualifiedTableNameToInfo, stateManager, emittedAt);
+      final List<AutoCloseableIterator<AirbyteMessage>> fullRefreshIterators =
+          getFullRefreshIterators(database, catalog, fullyQualifiedTableNameToInfo, stateManager, emittedAt);
+      final List<AutoCloseableIterator<AirbyteMessage>> iteratorList = Stream
+          .of(incrementalIterators, fullRefreshIterators)
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+
+      return AutoCloseableIterators
+          .appendOnClose(AutoCloseableIterators.concatWithEagerClose(iteratorList), () -> {
+            LOGGER.info("Closing database connection pool.");
+            Exceptions.toRuntime(this::close);
+            LOGGER.info("Closed database connection pool.");
+          });
+    } catch (final Exception exception) {
+      if (showEnhancedErrorDetails(exception)) {
+        AirbyteTraceMessageUtility.emitConfigErrorTrace(exception, exception.getMessage());
+      }
+      throw exception;
```

Comment:
Help me learn more.
In the above case we do this:

> return new AirbyteConnectionStatus()
>           .withStatus(Status.FAILED)
>           .withMessage(sshErrorMessage);

Why does this `throw` and that `return`? Maybe a difference between `check` and `read`?

Reply:
Done

Reply:
The ``check`` method expects an ``AirbyteConnectionStatus`` where the ``read`` method expects a combination of ``Record`` and ``State`` messages. Specifically for the ``read`` methods, unless an exception is thrown, the user does not see an error, so this leads to a silent failure. So, we need to throw an exception to show an error to the user. 

A future enhancement could be to change the protocol to handle additional message types - but I'm not sure what changes we'd need in the platform. Right now - it's not clear to me what benefit that would provide since we are just displaying an error message